### ISSUE TITLE
[HOLD for payment 2023-10-02] [HOLD] [Performance] ReportScreen unnecessary re-renders

### ISSUE THREAD

**Issue by mountiny** [Expensify/App#15163]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## What performance issue do we need to solve?

Right now, every time a report action updates the whole `<ReportActionsList />` will re-render. None of the items that get rendered are memoized. This means that on a single update (e.g. adding a reaction/editing the text) all mounted views of the list will need to re-render.

Lets investigate the ReportScreen component tree too as its quite render heavy

## What is the impact of this on end-users?

Re-rendering components when its contents haven't changed will in general slow down the app for no good reason

## List any benchmarks that show the severity of the issue

None yet

## Proposed solution (if any)

1. Memoize the <ReportActionItem /> - For this approach it would be good to have a version number for each report action, so we can quickly compare if we need to re-render it
2. Only pass an array of string ids to <ReportActionsList />, which represents the order of report actions. Then, <ReportActionItem /> just receives a string, and connects to onyx to get updates for the action from the store. This way when an update to a report happens (such as an added reaction), only the instance of <ReportActionItem /> has to re-render instead of the whole list
This approach might even be faster than the first option, as on an update not all items have to run their arePropsEqual function.

## List any benchmarks after implementing the changes to show impacts of the proposed solution (if any)

None yet

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** v1.2.71-1
**Reproducible in staging?:** yes
**Reproducible in production?:** yes
**Email or phone of affected tester (no customers):** N/A
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
**Expensify/Expensify Issue URL:** here
**Issue reported by:** @hannojg 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1675076963206029

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

**Comment by hannojg** [comment]
Please assign me for this one 👼

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days.  eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by mountiny** [comment]
I think this might one of @hannojg priorities once the Reactions go through some polish

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days. @hannojg eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by mountiny** [comment]
@hannojg How is this looking with your other projects? Reactions are almost done

**Comment by mountiny** [comment]
bumping this to Daily to make sure we investigate this as we are making this component even more robust with manual request so achieving great performance is important

**Comment by hannojg** [comment]
@mountiny I will pass this task on to @janicduplessis , while I will investigate the performance of other flows in the app.
Please assign him, the issue is in good hands with him 😊 

@janicduplessis Despite the title of the issue the goal here is to make the ReportScreen work just fast.

**Comment by mountiny** [comment]
@janicduplessis Can you please comment on this issue so I can assign you? thank you!

**Comment by janicduplessis** [comment]
Would you prefer multiple PRs with the different improvements separated or just a single one with all of the ReportScreen improvements?

**Comment by janicduplessis** [comment]
It also seems like the ReportActionComponent is now memoized so there isn't much excessive rendering going on anymore (https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionItem.js#L110-L123). I will further investigate if there are anything else expensive going on.

**Comment by mountiny** [comment]
@janicduplessis Thanks! I think the strategy depends if you can do parts of the refactorings separately and you think its easier to test and make sure there are not regressions then once big PR. But with big PR we can measure the performance gains a easier too/ more clearly.

So yeah I dont have a string preference without knowing how many options/ areas to improve there is

**Comment by mountiny** [comment]
LEts just merge that one PR and continue investigation for further improvements.

**Comment by mountiny** [comment]
Merged that one PR, thanks @janicduplessis! Any other findings?

@janicduplessis maybe you could test on this branch? https://github.com/Expensify/App/pull/18402 its using the new navigation architecture and there we can see the render times more problematic as the report screen is popped all the time on mobile going back to LHN and as we go back to the chat it needs to be rendered again showing the skeleton even though the data is locally available.

**Comment by janicduplessis** [comment]
@mountiny I haven’t had time to continue the investigation today, will do tomorrow. I will also have a look at that branch.

**Comment by mountiny** [comment]
thanks!

**Comment by mountiny** [comment]
I assume its bank holiday in canada today

**Comment by janicduplessis** [comment]
I started working on further improvements here https://github.com/Expensify/App/pull/19345. I will be available only on Thursday to keep working on this.

**Comment by mountiny** [comment]
Thanks!

**Comment by janicduplessis** [comment]
I found multiple improvements in https://github.com/Expensify/App/pull/19345, however one is blocked on updating our react-native-web fork since there is a bug currently with batchSize > 1 on web.

**Comment by mountiny** [comment]
Those are some great improvements

**Comment by mountiny** [comment]
@janicduplessis any updated here?

**Comment by mountiny** [comment]
Bumping

**Comment by mountiny** [comment]
Still waiting fro the RN web update.

**Comment by janicduplessis** [comment]
Still waiting for RN web update https://github.com/Expensify/App/issues/16660

**Comment by mountiny** [comment]
on hold

**Comment by mountiny** [comment]
Still on hold

**Comment by janicduplessis** [comment]
Yes, I will have another look towards end of week / next week

**Comment by mountiny** [comment]
Will be taken of hold soon

**Comment by mountiny** [comment]
@janicduplessis is making some progress on his PR

**Comment by janicduplessis** [comment]
I've been working with @ospfranco to address some issues in https://github.com/Expensify/App/pull/26772. This will cover fixes in my original PR except for the batch size one which still requires RN web update.

**Comment by mountiny** [comment]
@janicduplessis Could you please propose a problem/ solution for this issue is more specific based on your PR and we can link it to this tracking https://github.com/Expensify/App/issues/26087

**Comment by janicduplessis** [comment]
@mountiny I've focused efforts on https://github.com/Expensify/App/pull/26772 since most fixes in https://github.com/Expensify/App/pull/19345 no longer apply since there was many refactors since then. The only one that is still good is to increase batch size, which is on hold for RN web PR. Maybe I can change my PR to only include the batch size change and we can move tracking of render perf to https://github.com/Expensify/App/issues/26087.

**Comment by mountiny** [comment]
@janicduplessis I think that sounds great!

**Comment by janicduplessis** [comment]
Updated https://github.com/Expensify/App/pull/19345 to only include virtualized list batch size change.

**Comment by mountiny** [comment]
@janicduplessis to confirm, that does not need to be on hold for the RNW update, right?

**Comment by janicduplessis** [comment]
It still does, that was the one fix that needed the RNW update. If we want we could apply the fix only for native if we want to avoid the hold on RNW update.

**Comment by mountiny** [comment]
Hello @isabelastisser Can you please pay $500 to @sobitneupane for review and testing of this PR? Thanks!

**Comment by sobitneupane** [comment]
https://github.com/Expensify/App/issues/15163#issuecomment-1746707442

Requested payment on newDot

**Comment by mountiny** [comment]
Thanks then we can close this one

**Comment by JmillsExpensify** [comment]
$500 payment approved for @sobitneupane based on [this comment](https://github.com/Expensify/App/issues/15163#issuecomment-1746707442).
